blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/AbstractElasticIO.java

```diff
@@ -0,0 +1,463 @@
+
+    // todo: these should be instances per client.
+    protected final Timer searchTimer = Metrics.timer(getClass(), "Search Duration");
+    protected final Timer getNextTokensTimer = Metrics.timer(getClass(), "getNextTokens Duration");
+    protected final Timer writeTimer = Metrics.timer(getClass(), "Write Duration");
+    protected final Histogram batchHistogram = Metrics.histogram(getClass(), "Batch Sizes");
+    protected Meter classCastExceptionMeter = Metrics.meter(getClass(), "Failed Cast to IMetric");
+    protected Histogram queryBatchHistogram = Metrics.histogram(getClass(), "Query Batch Size");
+
+    public static String METRICS_TOKENS_AGGREGATE = "metric_tokens";
+    public static String ELASTICSEARCH_INDEX_NAME_WRITE = Configuration.getInstance().getStringProperty(ElasticIOConfig.ELASTICSEARCH_INDEX_NAME_WRITE);
+    public static String ELASTICSEARCH_INDEX_NAME_READ = Configuration.getInstance().getStringProperty(ElasticIOConfig.ELASTICSEARCH_INDEX_NAME_READ);
+
+    public static String ENUMS_INDEX_NAME_WRITE = Configuration.getInstance().getStringProperty(ElasticIOConfig.ELASTICSEARCH_ENUMS_INDEX_NAME_WRITE);
+    public static String ENUMS_INDEX_NAME_READ = Configuration.getInstance().getStringProperty(ElasticIOConfig.ELASTICSEARCH_ENUMS_INDEX_NAME_READ);
+
+    private final int LEVEL_0 = 0;
+    private final int LEVEL_1 = 1;
+    private final String REGEX_TOKEN_DELIMTER = "\\.";
+
+    public List<SearchResult> search(String tenant, String query) throws Exception {
+        return search(tenant, Arrays.asList(query));
+    }
+
+
+    public List<SearchResult> search(String tenant, List<String> queries) throws Exception {
+        String[] indexes = getIndexesToSearch();
+
+        return searchByIndexes(tenant, queries, indexes);
+    }
+
+    private List<SearchResult> searchByIndexes(String tenant, List<String> queries, String[] indexes) {
+        List<SearchResult> results = new ArrayList<SearchResult>();
+        Timer.Context multiSearchCtx = searchTimer.time();
+        queryBatchHistogram.update(queries.size());
+        BoolQueryBuilder bqb = boolQuery();
+        QueryBuilder qb;
+
+        for (String query : queries) {
+            GlobPattern pattern = new GlobPattern(query);
+            if (!pattern.hasWildcard()) {
+                qb = termQuery(ESFieldLabel.metric_name.name(), query);
+            } else {
+                qb = regexpQuery(ESFieldLabel.metric_name.name(), pattern.compiled().toString());
+            }
+
+            bqb.should(boolQuery()
+                            .must(termQuery(ESFieldLabel.tenantId.toString(), tenant))
+                            .must(qb)
+            );
+        }
+
+        SearchResponse response = client.prepareSearch(indexes)
+                .setRouting(tenant)
+                .setSize(100000)
+                .setVersion(true)
+                .setQuery(bqb)
+                .execute()
+                .actionGet();
+
+        multiSearchCtx.stop();
```

Comment:
This is existing code. I just refactored it. I will add try/finally blocks in the code I added which follow this pattern.